src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -1277,287 +804,149 @@ where
 
-    let mut batch_input = reclock_op.new_input_connection(
-        &batches,
-        Pipeline,
-        // We don't want frontier information to flow from the input to the
-        // output. The second input (the remap information) should be the stream
-        // that drives downstream timestamps.
-        vec![Antichain::new()],
-    );
     // Need to broadcast the remap changes to all workers.
     let remap_trace_updates = remap_trace_updates.broadcast();
     let mut remap_input = reclock_op.new_input(&remap_trace_updates, Pipeline);
 
-    reclock_op.build(move |mut capabilities| {
+    reclock_op.build(move |capabilities| async move {
         // The capability of the output after reclocking the source frontier
-        let mut cap_set = CapabilitySet::from_elem(
-            capabilities
-                .drain(..)
-                .exactly_one()
-                .expect("missing initial output capability"),
-        );
+        let mut cap_set = CapabilitySet::from_elem(capabilities.into_element());
 
-        let mut source_metrics =
-            SourceMetrics::new(&base_metrics, &name, id, &worker_id.to_string());
+        let mut source_metrics = SourceMetrics::new(&base_metrics, &name, id, &worker_id.to_string());
 
-        source_metrics
-            .resume_upper
-            .set(mz_persist_client::metrics::encode_ts_metric(&resume_upper));
+        source_metrics.resume_upper.set(mz_persist_client::metrics::encode_ts_metric(&resume_upper));
 
-        let mut batch_buffer = Vec::new();
         let mut remap_trace_buffer = Vec::new();
 
-        // The global view of the source_upper, which we track by combining
-        // summaries from the raw reader operators.
-        let mut global_source_upper = OffsetAntichain::new();
-
-        // Updates for `global_source_upper`. We stash these when they come in
-        // and process them once they are not beyond the frontier (in the domain
-        // of the timestamp that drives forward the ingestion pipeline) anymore.
-        let mut source_upper_updates = Vec::new();
+        let mut source_upper = MutableAntichain::new_bottom(FromTime::minimum());
 
         // Stash of batches that have not yet been timestamped.
-        let mut untimestamped_batches = VecDeque::new();
-
-        // We keep track of the lower of all stashed batches. We add the lower
-        // here when we receive a batch and stash it and we remove it when we
-        // have succesfully reclocked and emitted it.
-        //
-        // These batch lowers are an analog to "Capabilities" in timely. We use
-        // the lower to hold back on downgrading our output Capability, which we
-        // need when we can finally emit a reclocked batch.
-        let mut batch_capabilities = MutableOffsetAntichain::new();
+        let mut untimestamped_batches: Vec<Vec<(Result<SourceMessage<K, V>, SourceReaderError>, FromTime, D)>> = Vec::new();
```

Comment:
I don't think so because we receive owned `Vec<..>` from the handle and so keeping them like that does the minimal amount of work. Instead of copying all the contents and deallocating immediately we can skip the copying and defer the deallocation